syntax = "proto3";

package agent;

// Agent Looper service for continuous optimization
service AgentLooper {
  // Start optimization loop
  rpc StartLoop(StartRequest) returns (LoopResponse);

  // Stop optimization loop
  rpc StopLoop(StopRequest) returns (LoopResponse);

  // Get current status
  rpc GetStatus(StatusRequest) returns (StatusResponse);

  // Submit optimization goal
  rpc SubmitGoal(GoalRequest) returns (GoalResponse);

  // Request immediate analysis
  rpc Analyze(AnalyzeRequest) returns (AnalyzeResponse);

  // Approve/reject changes
  rpc ReviewChanges(ReviewRequest) returns (ReviewResponse);

  // Get metrics
  rpc GetMetrics(MetricsRequest) returns (MetricsResponse);

  // Stream log events
  rpc StreamLogs(LogRequest) returns (stream LogEvent);
}

// Messages
message StartRequest {
  string project_path = 1;
  repeated string goals = 2;
  bool auto_apply = 3;
  LooperConfig config = 4;
}

message StopRequest {
  string loop_id = 1;
}

message LoopResponse {
  bool success = 1;
  string loop_id = 2;
  string message = 3;
}

message StatusRequest {
  string loop_id = 1;
}

message StatusResponse {
  string loop_id = 1;
  bool running = 2;
  int32 iteration = 3;
  string current_phase = 4;
  repeated GoalStatus goals = 5;
  int64 started_at = 6;
  int64 last_activity = 7;
}

message GoalStatus {
  string name = 1;
  double progress = 2;
  bool completed = 3;
  string category = 4;
}

message GoalRequest {
  string loop_id = 1;
  Goal goal = 2;
}

message Goal {
  string name = 1;
  string description = 2;
  string category = 3;
  double target_value = 4;
  double current_value = 5;
  string unit = 6;
  map<string, string> metadata = 7;
}

message GoalResponse {
  bool success = 1;
  string message = 2;
  Goal goal = 3;
}

message AnalyzeRequest {
  string loop_id = 1;
  string focus_area = 2;  // e.g., "performance", "security", "features"
  map<string, string> context = 3;
}

message AnalyzeResponse {
  bool success = 1;
  string analysis = 2;
  repeated string recommendations = 3;
  repeated Issue issues = 4;
}

message Issue {
  string title = 1;
  string description = 2;
  string severity = 3;  // "low", "medium", "high", "critical"
  string category = 4;
  map<string, string> metadata = 5;
}

message ReviewRequest {
  string loop_id = 1;
  string iteration_id = 2;
  string decision = 3;  // "approve", "reject", "conditional"
  string feedback = 4;
}

message ReviewResponse {
  bool success = 1;
  string message = 2;
  bool applied = 3;
}

message MetricsRequest {
  string loop_id = 1;
  repeated string metric_names = 2;
}

message MetricsResponse {
  repeated Metric metrics = 1;
}

message Metric {
  string name = 1;
  double value = 2;
  string unit = 3;
  int64 timestamp = 4;
  map<string, string> labels = 5;
}

message LogRequest {
  string loop_id = 1;
  int64 since = 2;  // Unix timestamp
  LogLevel level = 3;
}

message LogEvent {
  int64 timestamp = 1;
  LogLevel level = 2;
  string message = 3;
  map<string, string> metadata = 4;
}

enum LogLevel {
  DEBUG = 0;
  INFO = 1;
  WARNING = 2;
  ERROR = 3;
  CRITICAL = 4;
}

message LooperConfig {
  int32 max_iterations = 1;
  int32 iteration_interval = 2;  // seconds
  bool auto_apply = 3;
  string model = 4;
  double temperature = 5;
  int32 max_tokens = 6;
}
